library/constant_time_impl.h

```diff
@@ -120,6 +148,34 @@ static inline mbedtls_ct_condition_t mbedtls_ct_bool(mbedtls_ct_uint_t x)
      * Otherwise, we define a plain C fallback which (in May 2023) does not get optimised into
      * conditional instructions or branches by trunk clang, gcc, or MSVC v19.
      */
+#if defined(MBEDTLS_CT_AARCH64_ASM) && (defined(MBEDTLS_CT_SIZE_32) || defined(MBEDTLS_CT_SIZE_64))
+    mbedtls_ct_uint_t s;
+    asm volatile ("neg %x[s], %x[x]                     \n\t"
+                  "orr %x[x], %x[s], %x[x]              \n\t"
+                  "asr %x[x], %x[x], 63"
+                  :
+                  [s] "=&r" (s),
+                  [x] "+&r" (x)
+                  :
+                  :
+                  );
+    return (mbedtls_ct_condition_t) x;
+#elif defined(MBEDTLS_CT_ARM_ASM) && defined(MBEDTLS_CT_SIZE_32)
+    uint32_t s;
+    asm volatile (".syntax unified                       \n\t"
+                  "negs %[s], %[x]                       \n\t"
+                  "orrs %[x], %[x], %[s]                 \n\t"
+                  "asrs %[x], %[x], #31                  \n\t"
+                  RESTORE_ASM_SYNTAX
+                  :
+                  [s] "=&l" (s),
+                  [x] "+&l" (x)
+                  :
+                  :
+                  "cc" /* clobbers flag bits */
+                  );
+    return (mbedtls_ct_condition_t) x;
+#else
     const mbedtls_ct_uint_t xo = mbedtls_ct_compiler_opaque(x);
```

Comment:
NOTE: Regarding the C implementation of this

```
    return (mbedtls_ct_condition_t) (((mbedtls_ct_int_t) ((-xo) | -(xo >> 1))) >>
                                     (MBEDTLS_CT_SIZE - 1));
```

The C standard says:

> The result of E1 >> E2 is E1 right-shifted E2 bit positions. If E1 has an unsigned type
> or if E1 has a signed type and a nonnegative value, the value of the result is the integral
> part of the quotient of E1 / 2E2. If E1 has a signed type and a negative value, the
> resulting value is implementation-defined.

Since we're intentionally propagating a set sign bit, we're in the "implementation-defined" category here :(

When compiling for host, we could have a static assert that `((mbedtls_ct_int_t)(-1) >> 1 == (mbedtls_ct_int_t)(-1))`, but "compiling for host" is pretty much the opposite of "embedded"

Reply:
Good spot. I've found a safe alternative that compiles to the same code (at least on aarch64, but presumably if the compiler sees the equivalence it will do the same elsewhere).